CONTRIBUTING.md

```diff
@@ -0,0 +1,408 @@
+
+[Git related](#git-related)
+ * [Our git Workflow](#our-git-workflow)
+ * [Writing good commit messages](#writing-good-commit-messages)
+ * [Creating, Reviewing and Merging Pull Requests](#creating-reviewing-and-merging-pull-requests)
+
+[Code Style](#code-style)
+ * [Automatic code formatting](#automatic-code-formatting)
+ * [Naming Scheme](#naming-scheme)
+ * [Whitespace](#whitespace)
+ * [Include statement grouping](#include-statement-grouping)
+ * [Other Code Formatting not handled by ClangFormat](#other-code-formatting-not-handled-by-clangformat)
+ * [CMake coding style](#cmake-coding-style)
+ 
+[Helper Scripts](#helper-scripts)
+ * [Create a new algorithm](#create-a-new-algorithm)
+ * [Converting CUDA code to HIP code](#converting-cuda-code-to-hip-code) 
+
+[Writing Tests](#writing-tests)
+ * [Testing know-how](#testing-know-how)
+ * [Some general rules](#some-general-rules)
+ * [Writing tests for kernels](#writing-tests-for-kernels)
+
+[Documentation style](#documentation-style)
+ * [Developer targeted notes](#developer-targeted-notes)
+ * [Whitespaces](#whitespaces)
+ * [Documenting examples](#documenting-examples)
+ 
+[Other general programming comments](#other-general-programming-comments)
+ * [C++ standard stream objects](#c++-standard-stream-objects)
+ * [Warnings](#warnings)
+ * [Avoiding circular dependencies](#avoiding-circular-dependencies)
+
+
+## Most important stuff (A TLDR)
+
+* `GINKGO_DEVEL_TOOLS` needs to be set to `on` to commit. This requires `clang-format` to be installed. See [Automatic code formatting](#automatic-code-formatting) for more details. Once installed, you can run `make format` in your `build/` folder to automatically format your modified files. Once formatted, you can commit your changes.
+
+* See [Our git workflow](#our-git-workflow) to get a quick overview of our workflow.
+
+* See [Creating, Reviewing and Merging Pull Requests](#creating-reviewing-and-merging-pull-requests) on how to create a Pull request.
+
+
+## Project structure
+
+Ginkgo is divided into a `core` module with common functionalities independent of the architecture, and several kernel modules (`reference`, `omp`, `cuda`, `hip`) which contain low-level computational routines for each supported architecture.
+
+### Extended header files
+
+Some header files from the core module have to be extended to include special functionality for specific architectures. An example of this is `core/base/math.hpp`, which has a GPU counterpart in `cuda/base/math.hpp`.
+For such files you should always include the version from the module you are working on, and this file will internally include its `core` counterpart.
+
+### Using library classes
+
+Creating new classes, it is allowed to use existing classes (polymorphic objects) inside the kernels for the distinct backends (reference/cuda/omp...). However,  it is not allowed to construct the kernels by creating new instances of a distinct (existing) class as this can result in compilation problems. 
```

Comment:
```suggestion
Creating new classes, it is allowed to use existing classes (polymorphic objects) inside the kernels for the distinct backends (reference/cuda/omp...). However,  it is not allowed to construct the kernels by creating new instances of a distinct (existing) class as this can result in linking problems due to circular dependencies between the `core` and backend library. 
```